tectonic, biber-for-tectonic: wrap tectonic with biber-2.17, fix #88067

bryango opened the pull request:
### Note
- This PR has undergone major design change. Please jump to https://github.com/NixOS/nixpkgs/pull/273740#issuecomment-1859932466 and onwards for the current design.
- It is easy to review this PR commit by commit: https://github.com/NixOS/nixpkgs/pull/273740/commits.

## Description of changes

The version of `biber` is strongly coupled to the TeX-like system it works with. For example, the biber version required by `tectonic` is not found in nixpkgs, since the texlive bundle used by tectonic lags behind the one in nixpkgs. This has been a [known](https://github.com/tectonic-typesetting/tectonic/issues/893) [issue](https://github.com/NixOS/nixpkgs/issues/88067). Previously, a user had to manually download (or pin nixpkgs to) a particular version of biber to use with the appropriate TeX system, in this case, `tectonic`.

In this PR, we introduce the `biber-for-tectonic` package as an alternative version of `biber` required by the current tectonic. At the same time, we make `tectonic` into a wrapper, which bundles `tectonic-unwrapped` with `biber-for-tectonic`. This fixes #88067.

Some nixpkgs code by @collares and @doronbehar is recycled for use in this PR.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bryango committed:
biber-for-tectonic: init at 2.17

The `tectonic` package depends on a version of `biber` that is generally
different from the one in the nixpkgs `texlive` bundle. This package
provides an override of biber suitable for use with tectonic.

For biber<=2.17 on perl>=5.36.0 a patch is needed.
This is recovered from a previous nixpkgs commit:

  https://github.com/NixOS/nixpkgs/commit/c784cdbf6b98e3e3aefb678b6ee8309cbb8bee15

Co-authored-by: Mauricio Collares <mauricio@collares.org>
Co-authored-by: Doron Behar <doron.behar@gmail.com>

bryango committed:
maintainers: add @bryango

Checked with:
  - nix eval --file maintainers/scripts/check-maintainers-sorted.nix
  - nix build --file lib/tests/maintainers.nix --print-out-paths --no-link

doronbehar reviewed (commented):
I'm not sure I like this approach... It is a bit too overwhelming IMO. Why not simply create a `biber2_17` attribute, that will `overrideAttrs` of `biber`, and we'll comment near it that tectonic is it's purpose?

I'm also not sure I want to wrap tectonic with biber, and I've expressed my opinion in https://github.com/NixOS/nixpkgs/commit/08899893970493ac0300780dcf3536b6bb881f77 . However I do tend to support wrapping tectonic with the correct `biber`, as upstream fails to update the bundle for a while now, and the issue is indeed annoying in general.

The downside of wrapping Tectonic though, is that Tectonic may break again in the future when the Tectonic's TeX bundle will be updated but no new Tectonic version will be released, because we will find out about it too late, and then again we'd have to force Tectonic to use the newer Biber.

I was thinking, perhaps we can force Tectonic to use a web bundle that we know for sure is compatible with a certain Biber package? Then at least we know we will overcome the downside presented above.

bryango commented:
Hi Doron! Thank you very much for the comment!

> I'm not sure I like this approach... It is a bit too overwhelming IMO. Why not simply create a `biber2_17` attribute, that will `overrideAttrs` of `biber`, and we'll comment near it that tectonic is it's purpose?

Yeah, that's perfectly fine! In fact, that's the first thing that I tried, but I later settled on the `biberVersions` abstraction because by doing this, when it comes to updating biber, everything can be done by updating `biber/versions.nix`, without touching any other file. We don't need to touch `all-packages.nix` or `biber` at all.

> The downside of wrapping Tectonic though, is that Tectonic may break again in the future when the Tectonic's TeX bundle will be updated but no new Tectonic version will be released, because we will find out about it too late, and then again we'd have to force Tectonic to use the newer Biber.

Yes, I agree and I share all your concerns. The reason I chose to do this now is that the [tectonic development is slowing down](https://github.com/tectonic-typesetting/tectonic/discussions/1122). This is bad news for tectonic but good news for downstream, since it would not be too difficult to keep up. Also since I am a keen user of both biber and tectonic, I have committed myself to be the maintainer (of the wrappers).

> I was thinking, perhaps we can force Tectonic to use a web bundle that we know for sure is compatible with a certain Biber package? Then at least we know we will overcome the downside presented above.

That's brilliant, but I am not sure whether there is an API for that in tectonic...

bryango commented:
I have simply the structure of `biberVersions` in `biber/version.nix`. Now there is no change of `biber/default.nix`.

However, we need to do `biber.override { version = ...; }` first and then `.overrideAttrs`, otherwise the `meta.name` attribute would be wrong. It seems that this is due to the wrapping of mkDerivation by `buildPerlModule`.

bryango commented on pkgs/tools/typesetting/biber/versions.nix:
```diff
@@ -0,0 +1,30 @@
+{
+  biber-from-texlive = biber;
+  biber-for-tectonic = let version = "2.17"; in (
+    biber.override {
+      ## this is necessary, otherwise wrong `meta.name`
+      texlive.pkgs.biber.texsource = {
+        inherit version;
+        inherit (biber) pname meta;
+      };
+    }
```
It is necessary to first override the `version` data here, which is passed to `buildPerlModule`, and then passed to `mkDerivation`. If we simply do `biber.overrideAttrs` the resulting package name will be wrong (`perl5.38.0-biber-2.19` instead of `2.17`), since it has already been preprocessed by `buildPerlModule`.

doronbehar reviewed (commented):
In addition to the below comments, if we'll settle down upon such a wrapper, maybe the `--web-bundle` option could allow us to stick to a specific bundle version which includes a biblatex version that supports our `biber` version we force here. Using this option needs testing though, and perhaps even documentation in the `meta.longDescription`.

doronbehar commented:
> but I later settled on the `biberVersions` abstraction because by doing this, when it comes to updating biber, everything can be done by updating `biber/versions.nix`, without touching any other file. We don't need to touch `all-packages.nix` or `biber` at all.

I agree that touching `all-packages.nix` is a headache :) Still you can create that derivation in a separate file, (use `pkgs/by-name/bi/biber-for-tectonic` to further ease yourself).

> > I was thinking, perhaps we can force Tectonic to use a web bundle that we know for sure is compatible with a certain Biber package? Then at least we know we will overcome the downside presented above.
> 
> That's brilliant, but I am not sure whether there is an API for that in tectonic...

See the main comment of the last review about the `--web-bundle` Tectonic option.

> Yes, I agree and I share all your concerns. The reason I chose to do this now is that the [tectonic development is slowing down](https://github.com/tectonic-typesetting/tectonic/discussions/1122). This is bad news for tectonic but good news for downstream, since it would not be too difficult to keep up. Also since I am a keen user of both biber and tectonic, I have committed myself to be the maintainer (of the wrappers).

Thanks for the link! I think it should be mentioned in the comments at the `biber-for-tectonic/package.nix` file that this thread strengthens our motivation to maintain `tectonic` as we do and that `biber-for-tectonic` package this way. In fact, if the `--web-bundle` option proves to work, perhaps we can even feel confident enough to force this wrapping for main tectonic as well, and not introduce a new `tectonic-with-biber` attribute.

There's also the middle-way betweeen your approach and the above, which would be to provide an unwrapped `tectonic.unwrapped` and keep the main `tectonic` derivation the wrapped one.

bryango commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,61 @@
+  };
+
+  ## produce the correct `meta.position` for `symlinkJoin`
+  pos = builtins.unsafeGetAttrPos "description" meta;
```
Wait a minute... I think now even without specifying `pos` the auto-generated `meta.position` is correct for `symlinkJoin` (was not like this a while ago). It seems that someone fixed this for trivial builders! I will remove this line then!

bryango commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,61 @@
+    makeBinaryWrapper "${lib.getBin biber}/bin/biber" \
+      $out/bin/biber-${biber.version}
+  '';
+  ## the biber executable is exposed as `biber-${biber.version}`
```
Oh, I think it would be better to expose biber in path; in my own experience, I would often run biber alone (without tectonic) to debug the bib entries. But if this is confusing, then I am okay to drop it. Please advise.

bryango commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,61 @@
+            pkgs = <...>; # import nixpkgs here
+          in pkgs.tectonic-with-biber.override {
+            biber = pkgs.biberVersions.biber-for-tectonic;
+          }
```
The original design is to have multiple bibers that the user can easily hot swap. I think we don't really need this anymore! Removed.

bryango commented on pkgs/tools/typesetting/biber/versions.nix:
```diff
@@ -0,0 +1,30 @@
+{ lib, callPackage, fetchFromGitHub, fetchpatch, biber }:
+
+{
+  biber-from-texlive = biber;
```
No problem, I have promoted `biber-from-texlive` to the top level.

bryango commented:
> maybe the `--web-bundle` option could allow us to stick to a specific bundle version which includes a biblatex version that supports our `biber` version we force here. Using this option needs testing though, and perhaps even documentation in the `meta.longDescription`.

> In fact, if the `--web-bundle` option proves to work, perhaps we can even feel confident enough to force this wrapping for main tectonic as well, and not introduce a new `tectonic-with-biber` attribute.

> There's also the middle-way betweeen your approach and the above, which would be to provide an unwrapped `tectonic.unwrapped` and keep the main `tectonic` derivation the wrapped one.

This is interesting, but I need some time to take a look and see how to implement that!

> I agree that touching `all-packages.nix` is a headache :) Still you can create that derivation in a separate file, (use `pkgs/by-name/bi/biber-for-tectonic` to further ease yourself).

This is cool! I didn't know this! But after some thought, I think this scheme places the biber override far away from `tectonic` and `biber`, which is somewhat cumbersome.

In the end, I decided to collapse `biberVersions` and move the `biber-for-tectonic` definition to `tectonic/biber.nix`. Please see whether it is okay this way!

bryango commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,61 @@
+
+  paths = [ tectonic ];
+  nativeBuildInputs = [ makeBinaryWrapper ];
+  passthru = { biber = biber; };
```
Wait sorry, I am a bit confused: shouldn't the `passthru` be the actual biber used, i.e. `biber-for-tectonic`?

doronbehar replied:
I meant:

```nix
passthru = {
  biber = biber-for-tectonic;
};
```

doronbehar commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,52 @@
+
+  paths = [ tectonic ];
+  nativeBuildInputs = [ makeBinaryWrapper ];
+  passthru = { inherit biber; };
```
This is wrong.

bryango replied:
Oh sorry, I was confused by your previous, previous comment. Now I understand and it makes sense. Fixed.

doronbehar commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,52 @@
+, symlinkJoin
+, tectonic
+, biber-for-tectonic
+, biber
```
You don't need `biber`, only `biber-for-tectonic`.

doronbehar commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,52 @@
+  postBuild = ''
+    wrapProgram $out/bin/tectonic \
+      --prefix PATH : "${lib.getBin biber-for-tectonic}/bin"
+    makeBinaryWrapper "${lib.getBin biber-for-tectonic}/bin/biber" \
+      $out/bin/biber-${biber-for-tectonic.version}
+  '';
+  # the biber executable is exposed as `biber-${biber.version}`
```
I'm sorry but I still don't understand this...

bryango replied:
Dropped for simplicity!

doronbehar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -25336,6 +25338,8 @@ with pkgs;
     harfbuzz = harfbuzzFull;
   };
 
+  tectonic-with-biber = callPackage ../tools/typesetting/tectonic/wrapper.nix { };
```
I still tend towards editing the regular `tectonic` expression, such that `tectonic.unwrapped` is the unwrapped tectonic, and `tectonic` is wrapped with `biber-for-tectonic`, using `symlinkJoin`...

bryango replied:
I am not sure about this, because biber is popular but not _that_ popular (unfortunately most journals still use bibtex). So it would be simpler to maintain a different package, for people that do need biber. But if you really think we should make the biber-wrapped tectonic as the default, I am also okay with that!

bryango commented:
There is currently [this upstream issue](https://github.com/tectonic-typesetting/tectonic/issues/1002) which prevents us to consistently override `web-bundle` for all cli interface. Currently, the default web bundle is hardcoded.

I am experimenting with patching tectonic to pin the hardcoded bundle; maybe that's too much? In any case, I haven't finished, but I will report back what I find.

doronbehar commented:
Hey @bryango sorry for the long waiting. Well done on https://github.com/tectonic-typesetting/tectonic/pull/1131 ! I really like the general idea, though I left a comment there as you probably noticed..

I think the best way to ensure reproducibility in Nixpkgs most cleanly would be to implement the following design:

- `tectonic-unwrapped` is a derivation that includes some version of your upstream patch https://github.com/tectonic-typesetting/tectonic/pull/1131 .
- `tectonic` is nothing but a wrapping derivation for `tectonic-unwrapped` that includes both:
- - `${biber-for-tectonic}/bin` in it's `$PATH`
- - The `$TECTONIC_WEB_BUNDLE` environment variable ensures the above `biber` is compatible with the bundle.
- `tectonic.passthru.unwrapped` could be another way to access `tectonic-unwrapped`.
- `tectonic.passthru.biber` could be another way to access `biber-for-tectonic`.

This way, users could use the attribute `tectonic-unwrapped` to play around with different bundles if they wish so, and be able to also play with the feature of your patch.

bryango commented:
Hey @doronbehar thank you for the comment! The structure you proposed looks good to me. I will wait a bit for upstream response and see what we should do. If there is no response, I will start to implement it based on our discussions!

doronbehar commented:
I'm afraid your work would be forgotten while we wait for upstream, and the issue will persist and most users won't get a proper fix as proposed here. We agree upon many of the changes and they aren't too entangled to the upstream PR IMO. Let's commit to Nixpkgs in the meantime the following commits:

- `maintainers: add bryango`
- `tectonic: add bryango as maintainer`
- `biber-for-tectonic: init at 2.17` # With you & I as maintainers
- `tectonic: redefine it as a wrapper derivation, using biber-for-tectonic`

In the last commit, add a comment that mentions your upstream PR, and that ideally the web bundle of the wrapped tectonic should use a pinned texlive bundle, to ensure reproducibility... I think you had a good phrased comment in the current PR that you can start with.

bryango committed:
tectonic: redefine to wrap it with `biber-for-tectonic`

The `tectonic` attribute is redefined to be a wrapper with a compatible
version of biber, provided by `biber-for-tectonic`.

The wrapper is partially recovered from a previous nixpkgs commit:

  https://github.com/NixOS/nixpkgs/commit/5aa8e9f0f90b0c9bbdf7b18ead18704d1622c509

Also:

- Remove unneeded makeBinaryWrapper input in `tectonic-unwrapped`.
- Add @bryango as a maintainer of both `tectonic-unwrapped` and
  `tectonic`.

Co-authored-by: Doron Behar <doron.behar@gmail.com>

bryango commented:
I have implemented the suggested changes!

Result of `nixpkgs-review pr 273740` run on x86_64-linux [(details)](https://github.com/bryango/nix-build-action/actions/runs/7204404228/job/19789583160)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>biber-for-tectonic</li>
    <li>biber-for-tectonic.devdoc</li>
    <li>tectonic</li>
    <li>tectonic-unwrapped</li>
  </ul>
</details>

doronbehar commented on pkgs/tools/typesetting/tectonic/default.nix:
```diff
@@ -23,8 +32,20 @@ rustPlatform.buildRustPackage rec {
+    # by specifying the environment variable `TECTONIC_WEB_BUNDLE_LOCKED`.
+    # Upstream PR: https://github.com/tectonic-typesetting/tectonic/pull/1131
+    # This patch should be removed once the upstream PR is merged
+    (fetchpatch {
```
What is the point in fetching the patch if you don't inject the `$TECTONIC_WEB_BUNDLE_LOCKED` variable? I'm a rather against this patch, but even if I do accept it, It doesn't make sense to me to add it here unconditionally, vs only in the main `tectonic` derivation. Note also that you anyway need to `overrideAttrs` since you rely upon build time environment variable.

doronbehar commented on pkgs/tools/typesetting/tectonic/tests.nix:
```diff
@@ -0,0 +1,66 @@
+  };
+  testfiles = "${biber-dev-source}/testfiles";
+
+  notice = builtins.toFile "tectonic-offline-notice" ''
+    # To fetch tectonic's web bundle, the tests require internet access,
+    # which is not available in a build sandbox. To run the tests, try:
+    # `nix-build --no-sandbox --attr tectonic.passthru.tests`
+  '';
```
So this test is only meant for manual running? Not by Hydra? I don't think so. How about adding a VM based test? These can access the internet. See examples at `nixos/tests/all-tests.nix`.

bryango replied:
I spent a night or two looking at the (sparsely documented) NixOS tests, and they are truly amazing beasts! However, it seems that although they do enjoy a network stack, they still do not have internet / "world wide web" access in a sandbox build. Fortunately, I do find out a way to achieve internet access: basically, we make the test as a "fetcher" like `fetchurl`. I have implemented this in #278410.

doronbehar commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,41 @@
+  passthru = {
+    unwrapped = tectonic-unwrapped;
+    biber = biber-for-tectonic;
+    tests = callPackage ./tests.nix { };
```
You can use a pattern such as done in `gotify-server` for the tests etc.

doronbehar reviewed (changes requested):
I really prefer `tectonic-unwrapped` to be a clean Tectonic without patches. I also dislike the fact that you have to recompile tectonic just to change the URL of the bundle - I dislike it being a build time option and not a runtime option, since it takes a lot of resources to compile it...

I'm sorry to put too many obstacles in the way of this PR, but I want the solution to eventually become something that upstream may approve. See my comment at https://github.com/tectonic-typesetting/tectonic/pull/1132#issuecomment-1864392520 .

bryango commented:
> I really prefer `tectonic-unwrapped` to be a clean Tectonic without patches. I also dislike the fact that you have to recompile tectonic just to change the URL of the bundle - I dislike it being a build time option and not a runtime option, since it takes a lot of resources to compile it...

I totally understand the desire to be perfect, but I am afraid I am running out of energy to continue... And I am getting more and more occupied in my day job :crying_cat_face: I will try to implement the remaining suggestions; the VM test is particularly interesting (I didn't know about that). For now, I will mark this PR as draft, but if you or anyone would like to build on the work here, feel free to do so!

doronbehar commented:
It's OK @bryango , you have done very good job anyway. I will checkout out this PR locally and do prepare to merge the things that we agree upon. I'll leave the upstream patches aside for now.

bryango commented:
> It's OK @bryango , you have done very good job anyway. I will checkout out this PR locally and do prepare to merge the things that we agree upon. I'll leave the upstream patches aside for now.

Thank you very much! I learned quite a lot along the way. If there is anything confusing in the code that you would like me to clarify, feel free to ping me! I might code less, but I can explain.

bryango commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,56 @@
+    # https://github.com/tectonic-typesetting/tectonic/discussions/1122
+    #
+    # Hence, we can be rather confident that for the near future, the online
+    # Tex bundle won't be updated and hence the biblatex distributed there
```
Nit: either `TeX` or `tex`. Sorry for my OCD...

bryango commented on pkgs/tools/typesetting/tectonic/wrapper.nix:
```diff
@@ -0,0 +1,56 @@
+  '';
+
+  meta = tectonic-unwrapped.meta // {
+    description = "Tectonic Tex/LaTeX engine, wrapped with a compatible biber";
```
Nit: same as above, either `TeX` or `tex`.

doronbehar replied:
> Nit: same as above, either `TeX` or `tex`.

Done (for both)!

bryango reviewed (commented):
I think the current wrapper is the best solution for now. Nits: I would like `Tex` to be `tex` or `TeX`. Otherwise, I think it's good to go! I am running `nixpkgs-review`.

Also, I think it would be easier for other people to review by starting a new PR and closing this one. What do you think? This PR can then be a place to track the downstream changes if the `--web-bundle` update finally works out in the upstream.

**Update:** Result of `nixpkgs-review pr 273740` run on x86_64-linux [(details)](https://github.com/bryango/nix-build-action/actions/runs/7204404228/job/19845995799)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>biber-for-tectonic</li>
    <li>biber-for-tectonic.devdoc</li>
    <li>tectonic</li>
    <li>tectonic-unwrapped</li>
  </ul>
</details>

doronbehar commented:
Will merge when CI is green.

doronbehar committed:
tectonic: symlink nextonic for all platforms

Probably this was conditioned on stdenv.isLinux by mistake.

doronbehar commented:
Fixed Darwin issue discovered by ofborg. Will wait for CI again..

doronbehar commented:
@ofborg build tectonic

bryango commented:
Wow awesome! Thank you very much for finishing this!
